src/lib/yacc/grammar.rs

```diff
@@ -702,6 +719,84 @@ fn nonterm_min_costs<F>(grm: &YaccGrammar, term_cost: F) -> Vec<u64>
     costs
 }
 
+
+/// Return the cost of the maximal string for each non-terminal in this grammar (u64::max_val()
+/// representing "this non-terminal can generate strings of infinite length"). The cost of a
+/// terminal is specified by the user-defined `term_cost` function.
+fn nonterm_max_costs<F>(grm: &YaccGrammar, term_cost: F) -> Vec<u64>
+               where F: Fn(TIdx) -> u64
+{
+    let mut done = vec![];
+    done.resize(grm.nonterms_len(), false);
+    let mut costs = vec![];
+    costs.resize(grm.nonterms_len(), 0);
+
+    // First mark all recursive non-terminals.
+    for i in 0..grm.nonterms_len() {
+        // Calling has_path so frequently is not exactly efficient...
+        if grm.has_path(NTIdx::from(i), NTIdx::from(i)) {
+            costs[i] = u64::max_value();
+            done[i] = true;
+        }
+    }
+
+    loop {
+        let mut all_done = true;
+        for i in 0..done.len() {
+            if done[i] {
+                continue;
+            }
+            all_done = false;
+            let mut hs_cmplt = None; // highest completed cost
+            let mut hs_noncmplt = None; // highest non-completed cost
+            'a: for p_idx in grm.nonterm_to_prods(NTIdx::from(i)).iter() {
+                let mut c: u64 = 0; // production cost
+                let mut cmplt = true;
+                for sym in grm.prod(*p_idx) {
+                    let sc = match *sym {
+                                 Symbol::Term(term_idx) => term_cost(term_idx),
+                                 Symbol::Nonterm(nt_idx) => {
+                                     if costs[usize::from(nt_idx)] == u64::max_value() {
+                                         // As soon as we find reference to an infinite
+                                         // non-terminal, we can stop looking.
+                                         hs_cmplt = Some(u64::max_value());
+                                         break 'a;
+                                     }
+                                     if !done[usize::from(nt_idx)] {
+                                         cmplt = false;
+                                     }
+                                     costs[usize::from(nt_idx)]
+                                 }
+                             };
+                    c = c.checked_add(sc).expect(
+                            "Overflow occurred when calculating non-term costs");
+                    if c == u64::max_value() {
```

Comment:
Can we reach this? It looks like `sc` can never be set to `u64::max_value` because of the break in Line 763.

Reply:
Unless we have a terminal that has a costs of exactly `u64::max_value` but then that wouldn't warrant a panic as that exact amount can be represented in `u64`.

Reply:
Ah wait. Does `c.checked_add(sc)` just set `maxint` when it overflows?

Reply:
No, it returns an Option: https://doc.rust-lang.org/std/primitive.u32.html#method.checked_add

Reply:
The easiest way to hit this is if a user sets the cost of a terminal to `u64::max_value()`.

Reply:
Can a user manually set this? But even if they do, then the panic message is wrong, because there could potentially be a terminal with a costs of maxint, right?

Reply:
Yep, the user could do that, but they'd be silly to do so because they're running the risk of overflow. The message just says "I can't deal with this" in effect. If anyone ever runs into this (which I doubt), I think we could reevaluate... but I'm not sure anyone ever will. This is just a super-safe sanity check.

Reply:
Fair enough.